In [4]:
# Importing essential libraries

# Data Gathering 
import pandas as pd
import numpy as np

# Data Visualization
import seaborn as sns



#### Data Gathering 

In [6]:
deliveries = pd.read_csv('deliveries.csv')
deliveries.sample(7)

match_id  inning                 batting_team           bowling_team  \
141768   1082612       1              Kings XI Punjab         Mumbai Indians   
24669     392230       1             Delhi Daredevils       Rajasthan Royals   
204819   1254104       2               Mumbai Indians    Chennai Super Kings   
86860     598041       1          Chennai Super Kings        Kings XI Punjab   
154077   1136576       1              Kings XI Punjab    Sunrisers Hyderabad   
150950   1136563       1  Royal Challengers Bangalore  Kolkata Knight Riders   
211899   1304062       2               Gujarat Titans           Punjab Kings   

        over  ball     batter     bowler   non_striker  batsman_runs  \
141768    18     4    HM Amla  JJ Bumrah      AR Patel             0   
24669      1     4  G Gambhir    A Singh      V Sehwag             0   
204819    11     3  SS Tiwary   DJ Bravo    KA Pollard             1   
86860     14     4   SK Raina  PP Chawla     JA Morkel             1   
154077    16     2    KK Nair  CJ Jordan      CH Gayle             1   
150950     1     1  Q de Kock  PP Chawla   BB McCullum             4   
211899    16     5  HH Pandya  RD Chahar  Shubman Gill             0   

        extra_runs  total_runs extras_type  is_wicket player_dismissed  \
141768           0           0         NaN          0              NaN   
24669            0           0         NaN          0              NaN   
204819           0           1         NaN          0              NaN   
86860            0           1         NaN          0              NaN   
154077           0           1         NaN          0              NaN   
150950           0           4         NaN          0              NaN   
211899           0           0         NaN          0              NaN   

       dismissal_kind fielder  
141768            NaN     NaN  
24669             NaN     NaN  
204819            NaN     NaN  
86860             NaN     NaN  
154077            NaN     NaN  
150950            NaN     NaN  
211899            NaN     NaN

`Column` | Description
:-|:-
`match_id` |	Unique identifier for each match, used to link with match details.
`inning`|	Indicates the innings (1 or 2) in the match.
`batting_team`|Name of the team batting in the given inning.
`bowling_team`|	Name of the team bowling in the given inning.
`over`|	The over number in the current inning (e.g., 1, 2, ..., 20).
`ball`|	The ball number within the over (1 to 6).
`batter`|	Name of the player currently batting.
`bowler`	|Name of the player currently bowling.
`non_striker`|	Name of the non-striker at the other end of the pitch.
`batsman_runs`|	Number of runs scored by the batsman on the current delivery (can include 0 for dot balls or 4/6 for boundaries).
`extra_runs`	|Number of extra runs (such as no-balls, wides) on this delivery.
`total_runs`	|Total runs scored on this ball (sum of batsman runs and extra runs).
`extras_type`	|Type of extra run, if applicable (e.g., wide, no-ball, bye, legbye).
`is_wicket`	|Indicates if the delivery resulted in a wicket (1 for wicket, 0 for no wicket).
`player_dismissed`|	Name of the player dismissed on the ball (if applicable, otherwise null).
`dismissal_kind`	|Type of dismissal (e.g., bowled, caught, run out, LBW).
`fielder`	|Name of the fielder involved in the dismissal (for dismissals like catch, run out, etc.; null for dismissals not involving a fielder).

In [8]:
matches = pd.read_csv('matches.csv')
matches.sample(7)

id   season       city        date match_type player_of_match  \
385    598062     2013    Chennai  2013-05-14     League        MS Dhoni   
370    598046     2013     Mumbai  2013-05-05     League      MG Johnson   
1000  1359525     2023  Ahmedabad  2023-05-07     League    Shubman Gill   
19     336001  2007/08    Chennai  2008-05-02     League        V Sehwag   
943   1304114     2022     Mumbai  2022-05-20     League        R Ashwin   
307    548367     2012  Bangalore  2012-05-14     League       AT Rayudu   
712   1175372     2019  Bengaluru  2019-04-05     League      AD Russell   

                                 venue                        team1  \
385    MA Chidambaram Stadium, Chepauk          Chennai Super Kings   
370                   Wankhede Stadium               Mumbai Indians   
1000  Narendra Modi Stadium, Ahmedabad               Gujarat Titans   
19     MA Chidambaram Stadium, Chepauk          Chennai Super Kings   
943          Brabourne Stadium, Mumbai          Chennai Super Kings   
307              M Chinnaswamy Stadium  Royal Challengers Bangalore   
712              M.Chinnaswamy Stadium  Royal Challengers Bangalore   

                      team2            toss_winner toss_decision  \
385        Delhi Daredevils    Chennai Super Kings           bat   
370     Chennai Super Kings         Mumbai Indians           bat   
1000   Lucknow Super Giants   Lucknow Super Giants         field   
19         Delhi Daredevils    Chennai Super Kings           bat   
943        Rajasthan Royals    Chennai Super Kings           bat   
307          Mumbai Indians         Mumbai Indians         field   
712   Kolkata Knight Riders  Kolkata Knight Riders         field   

                     winner   result  result_margin  target_runs  \
385     Chennai Super Kings     runs           33.0        169.0   
370          Mumbai Indians     runs           60.0        140.0   
1000         Gujarat Titans     runs           56.0        228.0   
19         Delhi Daredevils  wickets            8.0        170.0   
943        Rajasthan Royals  wickets            5.0        151.0   
307          Mumbai Indians  wickets            5.0        172.0   
712   Kolkata Knight Riders  wickets            5.0        206.0   

      target_overs super_over method          umpire1         umpire2  
385           20.0          N    NaN    C Shamshuddin       RJ Tucker  
370           20.0          N    NaN  HDPK Dharmasena       CK Nandan  
1000          20.0          N    NaN     AK Chaudhary  A Nand Kishore  
19            20.0          N    NaN        BF Bowden     K Hariharan  
943           20.0          N    NaN      CB Gaffaney   NA Patwardhan  
307           20.0          N    NaN            S Das     BR Doctrove  
712           20.0          N    NaN     AK Chaudhary     CB Gaffaney

`Column`	|Description
:-|:-
`id`	|Unique identifier for each match, similar to match_id in the ball-by-ball dataset, useful for linking datasets.
`season`|	Year in which the IPL season took place.
`city`	|City where the match was played.
`date`	|Date of the match.
`match_type`|	Type of match (could indicate regular season, playoff, or final).
`player_of_match`|	Name of the player awarded "Player of the Match."
`venue`|	Stadium where the match was played.
`team1`	|Name of the first team (order doesn’t imply batting or bowling first).
`team2`	|Name of the second team.
`toss_winner`|	Team that won the toss.
`toss_decision`|	Decision taken by the toss-winning team (bat or field).
`winner`|Name of the team that won the match.
`result`|	Outcome type of the match (e.g., normal, tie, no result).
`result_margin`|	Margin by which the team won (e.g., runs or wickets).
`target_runs`	|Target runs set for the chasing team (null if not applicable).
`target_overs`	|Target overs if the match is shortened (e.g., in rain-affected matches; null if not applicable).
`super_over`	|Indicates if the match had a super over (1 for yes, 0 for no).
`method`	|Method used to determine the result in rain-affected games (e.g., Duckworth-Lewis-Stern, or null if not applicable).
`umpire1`	|Name of the first on-field umpire.
`umpire2`	|Name of the second on-field umpire.

- here we have two datasets one containing ball by ball details and other containing match wise details.
- we are going to analyse each data set for insights.


#### Data Cleaning

In [12]:
deliveries.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 260920 entries, 0 to 260919
Data columns (total 17 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   match_id          260920 non-null  int64 
 1   inning            260920 non-null  int64 
 2   batting_team      260920 non-null  object
 3   bowling_team      260920 non-null  object
 4   over              260920 non-null  int64 
 5   ball              260920 non-null  int64 
 6   batter            260920 non-null  object
 7   bowler            260920 non-null  object
 8   non_striker       260920 non-null  object
 9   batsman_runs      260920 non-null  int64 
 10  extra_runs        260920 non-null  int64 
 11  total_runs        260920 non-null  int64 
 12  extras_type       14125 non-null   object
 13  is_wicket         260920 non-null  int64 
 14  player_dismissed  12950 non-null   object
 15  dismissal_kind    12950 non-null   object
 16  fielder           9354 non-null    obj

- here we get data information where we know basics about our data
    - There are 260920 rows and 17 columns in deliveries dataset.
    - some columns have missing values.


In [14]:
matches.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1095 entries, 0 to 1094
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               1095 non-null   int64  
 1   season           1095 non-null   object 
 2   city             1044 non-null   object 
 3   date             1095 non-null   object 
 4   match_type       1095 non-null   object 
 5   player_of_match  1090 non-null   object 
 6   venue            1095 non-null   object 
 7   team1            1095 non-null   object 
 8   team2            1095 non-null   object 
 9   toss_winner      1095 non-null   object 
 10  toss_decision    1095 non-null   object 
 11  winner           1090 non-null   object 
 12  result           1095 non-null   object 
 13  result_margin    1076 non-null   float64
 14  target_runs      1092 non-null   float64
 15  target_overs     1092 non-null   float64
 16  super_over       1095 non-null   object 
 17  method        

- Here  we have 1095 rows and 20 columns .
- there are some missing values in this data.
- Date column is in form of object which can be converted into date time for better analysis.

##### Further I cleaned both datas column wise 

#### Deliveries cleaning 

In [18]:
deliveries.sample(25)

match_id  inning                 batting_team  \
152740   1136570       2          Sunrisers Hyderabad   
116173    829767       2             Delhi Daredevils   
236767   1359519       1         Lucknow Super Giants   
47590     501224       1               Mumbai Indians   
205061   1254105       2               Delhi Capitals   
89082     598050       2        Kolkata Knight Riders   
41599     501198       1          Chennai Super Kings   
191266   1216542       1          Sunrisers Hyderabad   
136201    981017       1                Gujarat Lions   
38920     419154       2  Royal Challengers Bangalore   
102240    733995       1               Mumbai Indians   
104942    734017       2             Rajasthan Royals   
184153   1216513       1              Kings XI Punjab   
142553   1082615       2       Rising Pune Supergiant   
132692    980985       2              Kings XI Punjab   
13368     336040       2             Rajasthan Royals   
59666     548310       2             Delhi Daredevils   
47485     501223       2              Kings XI Punjab   
221150   1304100       2          Sunrisers Hyderabad   
243158   1370351       1               Mumbai Indians   
85144     598034       1          Chennai Super Kings   
230353   1359493       1          Sunrisers Hyderabad   
80557     598015       1                Pune Warriors   
136314    981017       2          Sunrisers Hyderabad   
235783   1359515       1          Chennai Super Kings   

                       bowling_team  over  ball            batter  \
152740        Kolkata Knight Riders    13     3   Shakib Al Hasan   
116173              Kings XI Punjab    11     3           SS Iyer   
236767          Chennai Super Kings     2     6         KR Mayers   
47590               Deccan Chargers    12     3         RG Sharma   
205061          Sunrisers Hyderabad    10     1           SS Iyer   
89082                Mumbai Indians    13     1            DB Das   
41599         Kolkata Knight Riders     1     1        S Anirudha   
191266              Kings XI Punjab    13     4       JM Bairstow   
136201          Sunrisers Hyderabad    15     3         RA Jadeja   
38920              Rajasthan Royals     8     2        RV Uthappa   
102240          Chennai Super Kings    18     5           AP Tare   
104942               Mumbai Indians     2     2         UBT Chand   
184153          Chennai Super Kings     2     2          KL Rahul   
142553          Sunrisers Hyderabad    11     2          MS Dhoni   
132692               Mumbai Indians    14     2           M Vijay   
13368           Chennai Super Kings     0     6          NK Patel   
59666   Royal Challengers Bangalore    11     4        GJ Maxwell   
47485              Delhi Daredevils    15     3          SE Marsh   
221150  Royal Challengers Bangalore    18     6  Fazalhaq Farooqi   
243158         Lucknow Super Giants     4     4          SA Yadav   
85144         Kolkata Knight Riders    13     5        MEK Hussey   
230353        Kolkata Knight Riders     3     5          HC Brook   
80557           Chennai Super Kings    13     6       LRPL Taylor   
136314                Gujarat Lions    13     5           NV Ojha   
235783                 Punjab Kings     5     6        RD Gaikwad   

                  bowler      non_striker  batsman_runs  extra_runs  \
152740     Kuldeep Yadav    KS Williamson             1           0   
116173       NLTC Perera       MA Agarwal             4           0   
236767         DL Chahar          M Vohra             1           0   
47590            PP Ojha        A Symonds             4           0   
205061       Rashid Khan         S Dhawan             1           0   
89082    Harbhajan Singh        R McLaren             6           0   
41599          YK Pathan         SK Raina             1           0   
191266  Mujeeb Ur Rahman        DA Warner             6           0   
136201          TA Boult         AJ Finch             1           0   
38920           SK W

#### match_id

In [20]:
deliveries['match_id'].isnull().sum()

0

In [21]:
deliveries['match_id'].nunique()

1095

- In match_id column there isn't any null value .
- Their are 1095 unique match id's which matches #rows in matches dataset.
   i.e. We have information of all matches.

#### inning

In [24]:
deliveries['inning'].isnull().sum()

0

In [25]:
deliveries['inning'].unique()

array([1, 2, 3, 4, 5, 6], dtype=int64)

- here data does not have null values.
- but innings values are little disturbed, as we know their should be only 2 innings but have till 6.
    - 3rd and 4th inning is their due to a super over but 5th and 6th inning is a entry error.

In [27]:
# If inning are 5 changing to 3 and if 6 changed to 4
def inning_replacing(inning):
    if inning == 5:
        return 3
    elif inning == 6:
        return 4
    else:
        return inning

In [28]:
deliveries['inning']=deliveries['inning'].apply(inning_replacing)

In [29]:
deliveries['inning'].unique()

array([1, 2, 3, 4], dtype=int64)

#### batting team

In [31]:
deliveries['batting_team'].isnull().sum()

0

In [32]:
deliveries['batting_team'].nunique()

19

In [33]:
deliveries['batting_team'].unique()

array(['Kolkata Knight Riders', 'Royal Challengers Bangalore',
       'Chennai Super Kings', 'Kings XI Punjab', 'Rajasthan Royals',
       'Delhi Daredevils', 'Mumbai Indians', 'Deccan Chargers',
       'Kochi Tuskers Kerala', 'Pune Warriors', 'Sunrisers Hyderabad',
       'Rising Pune Supergiants', 'Gujarat Lions',
       'Rising Pune Supergiant', 'Delhi Capitals', 'Punjab Kings',
       'Lucknow Super Giants', 'Gujarat Titans',
       'Royal Challengers Bengaluru'], dtype=object)

- their is no null value in batting team
- over the years names of teams are changed so same team is present with several names or their are some speeling difference

In [35]:
# correcting teams name if differs
def change_team_name(team_name):
    if team_name in ['Rising Pune Supergiants','Pune Warriors']:
        return 'Rising Pune Supergiant'
    elif team_name == 'Royal Challengers Bengaluru':
        return 'Royal Challengers Bangalore'
    elif team_name == 'Deccan Chargers':
        return 'Sunrisers Hyderabad'
    elif team_name == 'Delhi Daredevils':
        return 'Delhi Capitals'
    elif team_name == 'Gujarat Lions':
        return 'Gujarat Titans'
    elif team_name == 'Kings XI Punjab':
        return 'Punjab Kings'
    else:
        return team_name

In [36]:
deliveries['batting_team'] = deliveries['batting_team'].apply(change_team_name)

In [37]:
deliveries['batting_team'].nunique()

12

- there were 19 teams in batting team column which reduced to 12 as we know till today 12 teams were played in IPL

#### bowling team

In [40]:
deliveries['bowling_team'].isnull().sum()

0

In [41]:
deliveries['bowling_team'].nunique()

19

In [42]:
deliveries['bowling_team'] = deliveries['bowling_team'].apply(change_team_name)

In [43]:
deliveries['bowling_team'].nunique()

12

- their were 19 teams reduced to 12 with same problem as batting team

#### Over

In [46]:
deliveries['over'].isnull().sum()

0

In [47]:
deliveries['over'].unique()

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19], dtype=int64)

- 20 overs are present 1 over is calculated after completing.

#### ball

In [50]:
deliveries['ball'].isnull().sum()

0

In [51]:
deliveries['ball'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11], dtype=int64)

- as we know each over has 6 balls their can be wides and no balls in some overs.

#### batter

In [54]:
deliveries['batter'].isnull().sum()

0

In [55]:
deliveries['batter'].nunique()

673

In [56]:
deliveries['batter'].unique()

array(['SC Ganguly', 'BB McCullum', 'RT Ponting', 'DJ Hussey',
       'Mohammad Hafeez', 'R Dravid', 'W Jaffer', 'V Kohli', 'JH Kallis',
       'CL White', 'MV Boucher', 'B Akhil', 'AA Noffke', 'P Kumar',
       'Z Khan', 'SB Joshi', 'PA Patel', 'ML Hayden', 'MEK Hussey',
       'MS Dhoni', 'SK Raina', 'JDP Oram', 'S Badrinath', 'K Goel',
       'JR Hopes', 'KC Sangakkara', 'Yuvraj Singh', 'SM Katich',
       'IK Pathan', 'T Kohli', 'YK Pathan', 'SR Watson', 'M Kaif',
       'DS Lehmann', 'RA Jadeja', 'M Rawat', 'D Salunkhe', 'SK Warne',
       'SK Trivedi', 'G Gambhir', 'V Sehwag', 'S Dhawan', 'L Ronchi',
       'ST Jayasuriya', 'DJ Thornely', 'RV Uthappa', 'PR Shah',
       'AM Nayar', 'SM Pollock', 'Harbhajan Singh', 'S Chanderpaul',
       'LRPL Taylor', 'AC Gilchrist', 'Y Venugopal Rao', 'VVS Laxman',
       'A Symonds', 'RG Sharma', 'SB Styris', 'AS Yadav', 'SB Bangar',
       'WPUJC Vaas', 'RP Singh', 'WP Saha', 'LR Shukla',
       'DPMD Jayawardene', 'S Sohal', 'B Lee', 'PP Cha

- there are total 673 batter played in IPL till now

#### bowler

In [59]:
deliveries['bowler'].isnull().sum()

0

In [60]:
deliveries['bowler'].nunique()

530

In [61]:
deliveries['bowler'].unique()

array(['P Kumar', 'Z Khan', 'AA Noffke', 'JH Kallis', 'SB Joshi',
       'CL White', 'AB Dinda', 'I Sharma', 'AB Agarkar', 'SC Ganguly',
       'LR Shukla', 'B Lee', 'S Sreesanth', 'JR Hopes', 'IK Pathan',
       'K Goel', 'PP Chawla', 'WA Mota', 'JDP Oram', 'MS Gony',
       'M Muralitharan', 'P Amarnath', 'Joginder Sharma', 'GD McGrath',
       'B Geeves', 'MF Maharoof', 'R Bhatia', 'DL Vettori', 'MM Patel',
       'SR Watson', 'SK Trivedi', 'SK Warne', 'YK Pathan', 'D Salunkhe',
       'R Vinay Kumar', 'B Akhil', 'A Nehra', 'SM Pollock', 'DS Kulkarni',
       'ST Jayasuriya', 'Harbhajan Singh', 'AM Nayar', 'M Kartik',
       'Mohammad Hafeez', 'DJ Hussey', 'WPUJC Vaas', 'RP Singh',
       'SB Styris', 'SB Bangar', 'A Symonds', 'PP Ojha', 'Pankaj Singh',
       'Mohammad Asif', 'VY Mahesh', 'Shahid Afridi', 'DJ Bravo',
       'VS Yeligati', 'MA Khote', 'D Kalyankrishna', 'VRV Singh',
       'Sohail Tanvir', 'A Kumble', 'DNT Zoysa', 'SD Chitnis',
       'Yuvraj Singh', 'Shoaib Malik',

- 530 bowlers played IPL till now

#### non-striker

In [64]:
deliveries['non_striker'].isnull().sum()

0

In [65]:
deliveries['non_striker'].nunique()

663

In [66]:
deliveries['non_striker'].unique()

array(['BB McCullum', 'SC Ganguly', 'RT Ponting', 'DJ Hussey',
       'Mohammad Hafeez', 'W Jaffer', 'R Dravid', 'V Kohli', 'JH Kallis',
       'CL White', 'MV Boucher', 'B Akhil', 'AA Noffke', 'P Kumar',
       'Z Khan', 'SB Joshi', 'ML Hayden', 'PA Patel', 'MEK Hussey',
       'MS Dhoni', 'SK Raina', 'JDP Oram', 'S Badrinath', 'JR Hopes',
       'K Goel', 'KC Sangakkara', 'Yuvraj Singh', 'SM Katich',
       'IK Pathan', 'YK Pathan', 'T Kohli', 'SR Watson', 'M Kaif',
       'DS Lehmann', 'RA Jadeja', 'M Rawat', 'D Salunkhe', 'SK Warne',
       'SK Trivedi', 'V Sehwag', 'G Gambhir', 'S Dhawan', 'ST Jayasuriya',
       'L Ronchi', 'DJ Thornely', 'RV Uthappa', 'PR Shah', 'AM Nayar',
       'SM Pollock', 'Harbhajan Singh', 'S Chanderpaul', 'LRPL Taylor',
       'Y Venugopal Rao', 'AC Gilchrist', 'VVS Laxman', 'A Symonds',
       'SB Styris', 'AS Yadav', 'SB Bangar', 'WPUJC Vaas', 'RP Singh',
       'PP Ojha', 'WP Saha', 'LR Shukla', 'DPMD Jayawardene', 'S Sohal',
       'PP Chawla', 'WA M

- there are total 663 unique values in non-striker

#### batsmam_runs

In [69]:
deliveries['batsman_runs'].isnull().sum()

0

In [70]:
deliveries['batsman_runs'].unique()

array([0, 4, 6, 1, 2, 5, 3], dtype=int64)

- values are till 6 that a batter can strike

#### extra_runs

In [73]:
deliveries['extra_runs'].isnull().sum()

0

In [74]:
deliveries['extra_runs'].unique()

array([1, 0, 5, 4, 2, 3, 7], dtype=int64)

- because of bowlers mistake some penalties are gives so the value is 7

#### Total_runs

In [77]:
deliveries['total_runs'].isnull().sum()

0

In [78]:
deliveries['total_runs'].unique()

array([1, 0, 4, 6, 5, 2, 3, 7], dtype=int64)

- there is no anamolous behaviour in this column

#### extras_type

In [81]:
deliveries['extras_type'].isnull().sum()

246795

In [82]:
deliveries['extras_type'].unique()

array(['legbyes', nan, 'wides', 'byes', 'noballs', 'penalty'],
      dtype=object)

- here we have 246795 null values which is valid as their isnt any extra runs for these ball.
- there is no need to fill these null values 
- we can futher drop this column as it is of no use.

#### is_wicket

In [85]:
deliveries['is_wicket'].isnull().sum()

0

In [86]:
deliveries['is_wicket'].unique()

array([0, 1], dtype=int64)

- in this column 0 means no wicket and 1 mean they caught a wicket.


#### player_dismissed

In [89]:
deliveries['player_dismissed'].isnull().sum()

247970

In [90]:
deliveries['player_dismissed'].unique()

array([nan, 'SC Ganguly', 'RT Ponting', 'DJ Hussey', 'R Dravid',
       'V Kohli', 'JH Kallis', 'W Jaffer', 'MV Boucher', 'B Akhil',
       'CL White', 'AA Noffke', 'Z Khan', 'SB Joshi', 'PA Patel',
       'ML Hayden', 'MS Dhoni', 'SK Raina', 'JDP Oram', 'K Goel',
       'JR Hopes', 'Yuvraj Singh', 'KC Sangakkara', 'T Kohli',
       'YK Pathan', 'SR Watson', 'DS Lehmann', 'M Kaif', 'M Rawat',
       'RA Jadeja', 'SK Warne', 'V Sehwag', 'L Ronchi', 'DJ Thornely',
       'ST Jayasuriya', 'PR Shah', 'RV Uthappa', 'AM Nayar', 'SM Pollock',
       'S Chanderpaul', 'LRPL Taylor', 'Y Venugopal Rao', 'VVS Laxman',
       'AC Gilchrist', 'RG Sharma', 'SB Styris', 'AS Yadav', 'A Symonds',
       'WPUJC Vaas', 'SB Bangar', 'PP Ojha', 'BB McCullum', 'WP Saha',
       'Mohammad Hafeez', 'DPMD Jayawardene', 'IK Pathan', 'B Lee',
       'S Sohal', 'Kamran Akmal', 'Shahid Afridi', 'G Gambhir',
       'MEK Hussey', 'DJ Bravo', 'MA Khote', 'Harbhajan Singh',
       'GC Smith', 'D Salunkhe', 'RR Sarwan',

- in this column we have the player of whose the wicket is caught
- the data is of every ball so their are many null value which is valid. 
- the column does not gives any specific value so we might delete it later.

#### dismissal_kind

In [93]:
deliveries['dismissal_kind'].isnull().sum()

247970

In [94]:
deliveries['dismissal_kind'].unique()

array([nan, 'caught', 'bowled', 'run out', 'lbw', 'retired hurt',
       'stumped', 'caught and bowled', 'hit wicket',
       'obstructing the field', 'retired out'], dtype=object)

- here also null values are valid due to same reason

#### fielder

In [97]:
deliveries['fielder'].isnull().sum()

251566

In [98]:
deliveries['fielder'].unique()

array([nan, 'JH Kallis', 'P Kumar', 'CL White', 'M Kartik', 'RT Ponting',
       'WP Saha', 'AB Agarkar', 'BB McCullum', 'KC Sangakkara',
       'IK Pathan', 'PA Patel', 'Joginder Sharma', 'P Amarnath',
       'M Manhas', 'B Geeves', 'V Kohli', 'Z Khan', 'MV Boucher',
       'AM Nayar', 'L Ronchi', 'LR Shukla', 'Mohammad Hafeez', 'AB Dinda',
       'DJ Hussey', 'SC Ganguly', 'RG Sharma', 'SB Styris', 'A Symonds',
       'Kamran Akmal', 'RA Jadeja', 'PP Chawla', 'S Dhawan', 'KD Karthik',
       'MK Tiwary', 'GD McGrath', 'G Gambhir', 'AC Gilchrist', 'DJ Bravo',
       'Harbhajan Singh', 'S Badrinath', 'JDP Oram', 'SK Raina',
       'D Salunkhe', 'SB Bangar', 'D Kalyankrishna', 'VVS Laxman',
       'DS Kulkarni', 'AM Rahane', 'A Nehra', 'B Lee', 'RR Sarwan',
       'Yuvraj Singh', 'MM Patel', 'GC Smith', 'M Rawat', 'Sohail Tanvir',
       'A Kumble', 'M Muralitharan', 'MS Dhoni', 'I Sharma', 'PP Ojha',
       'DPMD Jayawardene', 'V Sehwag', 'W Jaffer', 'R Vinay Kumar',
       'R Dravid',

- fielders who caused the dismissal of the player
- therefore null values are valid.

#### checking duplicate rows

In [156]:
deliveries[deliveries.duplicated()]

Empty DataFrame
Columns: [match_id, inning, batting_team, bowling_team, over, ball, batter, bowler, non_striker, batsman_runs, extra_runs, total_runs, extras_type, is_wicket, player_dismissed, dismissal_kind, fielder]
Index: []

In [ ]:
- their isn't any d